File: src/main/java/com/radixpro/enigma/statistics/ui/ScenarioDetails.java
/*
 * Jan Kampherbeek, (c) 2020.
 * Enigma is open source.
 * Please check the file copyright.txt in the root of the source for further details.
 *
 */

package com.radixpro.enigma.statistics.ui;

import com.radixpro.enigma.Rosetta;
import com.radixpro.enigma.statistics.ui.domain.ScenarioFe;
import com.radixpro.enigma.statistics.ui.helpers.ScenDetailsText;
import com.radixpro.enigma.ui.creators.*;
import javafx.scene.Scene;
import javafx.scene.control.Button;
import javafx.scene.control.ButtonBar;
import javafx.scene.control.Label;
import javafx.scene.layout.Pane;
import javafx.scene.layout.VBox;
import javafx.stage.Stage;
import org.jetbrains.annotations.NotNull;

import static com.radixpro.enigma.ui.shared.UiDictionary.GAP;
import static com.radixpro.enigma.ui.shared.UiDictionary.TITLE_HEIGHT;

public class ScenarioDetails {

   private static final double HEIGHT = 500.0;
   private static final double HEIGHT_DETAILS = 300.0;
   private static final double WIDTH = 600.0;
   private final StatsFacade facade;
   private final ScenDetailsText scenDetailsText;
   private Stage stage;
   private Pane paneTitle;
   private String projName;
   private String scenName;

   public ScenarioDetails(@NotNull final StatsFacade facade, @NotNull final ScenDetailsText scenDetailsText) {
      this.facade = facade;
      this.scenDetailsText = scenDetailsText;
   }

   public void show(@NotNull String scenName, @NotNull final String projName) {
      this.projName = projName;
      this.scenName = scenName;
      stage = new Stage();
      initialize();
      stage.setTitle(Rosetta.getText("ui.stats.scendetail.title"));
      stage.setScene(new Scene(createVBox()));
      stage.showAndWait();
   }

   private void initialize() {
      Label lblTitle = new LabelBuilderObs("ui.stats.scendetail.title").setPrefWidth(WIDTH).setStyleClass("titletext").build();
      paneTitle = new PaneBuilder().setWidth(WIDTH).setHeight(TITLE_HEIGHT).setWidth(WIDTH).setStyleClass("titlepane").setChildren(lblTitle).build();
   }


   private VBox createVBox() {
      return new VBoxBuilder().setHeight(HEIGHT).setWidth(WIDTH).setPadding(GAP).setChildren(
            paneTitle,
            createPaneDetails(),
            createPaneButtonBar()
      ).build();
   }

   private Pane createPaneDetails() {
      ScenarioFe scenario = facade.readScenario(scenName, projName);
      String detailText = scenDetailsText.createText(scenario);
      Label lblDetails = new LabelBuilderObs("").setText(detailText).setPrefWidth(WIDTH).build();
      lblDetails.setWrapText(true);
      return new PaneBuilder().setHeight(HEIGHT_DETAILS).setWidth(WIDTH).setChildren(lblDetails).build();
   }

   private Pane createPaneButtonBar() {
      Button btnHelp = new ButtonBuilderObs("ui.shared.btn.help").setDisabled(false).setFocusTraversable(true).build();
      btnHelp.setOnAction(e -> onHelp());
      Button btnClose = new ButtonBuilderObs("ui.shared.btn.exit").setDisabled(false).setFocusTraversable(true).build();
      btnClose.setOnAction(e -> stage.close());
      ButtonBar buttonBar = new ButtonBarBuilder().setButtons(btnHelp, btnClose).build();
      return new PaneBuilder().setWidth(WIDTH).setHeight(30.0).setChildren(buttonBar).build();
   }

   private void onHelp() {
      // todo create help
   }

}

File: src/main/java/com/radixpro/enigma/statistics/ui/ScenarioDetails.kt
/*
 * Jan Kampherbeek, (c) 2020.
 * Enigma is open source.
 * Please check the file copyright.txt in the root of the source for further details.
 *
 */
package com.radixpro.enigma.statistics.ui

import com.radixpro.enigma.Rosetta
import com.radixpro.enigma.Rosetta.getText
import com.radixpro.enigma.statistics.ui.helpers.ScenDetailsText
import com.radixpro.enigma.ui.creators.*
import com.radixpro.enigma.ui.shared.Help
import com.radixpro.enigma.ui.shared.UiDictionary
import javafx.event.EventHandler
import javafx.scene.Scene
import javafx.scene.control.Button
import javafx.scene.layout.Pane
import javafx.scene.layout.VBox
import javafx.stage.Stage

class ScenarioDetails(
    private val facade: StatsFacade,
    private val scenDetailsText: ScenDetailsText
) {
    // texts
    private lateinit var projName: String
    private lateinit var scenName: String
    private lateinit var txtBtnClose: String
    private lateinit var txtBtnDelete: String
    private lateinit var txtBtnHelp: String
    private lateinit var txtTitle: String

    // buttons
    private lateinit var btnClose: Button
    private lateinit var btnDelete: Button
    private lateinit var btnHelp: Button

    private lateinit var stage: Stage


    fun show(scenName: String, projName: String) {
        this.projName = projName
        this.scenName = scenName
        stage = Stage()
        initialize()
        stage.title = txtTitle
        stage.scene = Scene(createVBox())
        stage.showAndWait()
    }

    private fun initialize() {
        defineTexts()
        defineButtons()
    }

    private fun defineTexts() {
        txtBtnClose = getText("ui.shared.btn.exit")
        txtBtnDelete = getText("ui.shared.btn.delete")
        txtBtnHelp = getText("ui.shared.btn.help")
        txtTitle = getText("ui.stats.scendetail.title")
    }

    private fun defineButtons() {
        btnClose = ButtonBuilder(txtBtnClose).setDisabled(false).setFocusTraversable(true).build()
        btnDelete = ButtonBuilder(txtBtnDelete).setDisabled(false).setFocusTraversable(true).build()
        btnHelp = ButtonBuilder(txtBtnHelp).setDisabled(false).setFocusTraversable(true).build()
        btnClose.onAction = EventHandler { onClose() }
        btnDelete.onAction = EventHandler { onDelete() }
        btnHelp.onAction = EventHandler { onHelp() }
    }

    private fun createVBox(): VBox {
        return VBoxBuilder().setHeight(HEIGHT).setWidth(WIDTH).setPadding(UiDictionary.GAP).setChildren(
            createPaneTitle(),
            createPaneDetails(),
            createPaneButtonBar()
        ).build()
    }

    private fun createPaneTitle(): Pane {
        val label = LabelBuilder(txtTitle).setPrefWidth(WIDTH).setPrefHeight(UiDictionary.TITLE_HEIGHT).setStyleClass("titletext").build()
        return PaneBuilder().setWidth(UiDictionary.INPUT_WIDTH).setHeight(UiDictionary.TITLE_HEIGHT).setStyleClass("titlepane").setChildren(label).build()
    }

    private fun createPaneDetails(): Pane {
        val scenario = facade.readScenario(scenName, projName)
        val detailText = scenDetailsText.createText(scenario)
        val lblDetails = LabelBuilderObs("").setText(detailText).setPrefWidth(WIDTH).build()
        lblDetails.isWrapText = true
        return PaneBuilder().setHeight(HEIGHT_DETAILS).setWidth(WIDTH).setChildren(lblDetails).build()
    }

    private fun createPaneButtonBar(): Pane {
        val buttonBar = ButtonBarBuilder().setButtons(btnDelete, btnHelp, btnClose).build()
        return PaneBuilder().setWidth(WIDTH).setHeight(30.0).setChildren(buttonBar).build()
    }

    private fun onDelete() {
        facade.deleteScenario(scenName, projName)
    }

    private fun onHelp() {
        Help(Rosetta.getHelpText("help.statsscendetails.title"), Rosetta.getHelpText("help.statsscendetails.content"))
    }

    private fun onClose() {
        stage.close()
    }

    companion object {
        private const val HEIGHT = 500.0
        private const val HEIGHT_DETAILS = 300.0
        private const val WIDTH = 600.0
    }
}